### ISSUE TITLE
Package installation fails with files in top-level of archive

### ISSUE THREAD

**Issue by adamwallred** [Homebrew/brew#4509]
- [ X ] are reporting a bug others will be able to reproduce and not asking a question. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh
- [ X ] ran a `brew` command and reproduced the problem with multiple formulae? If it's a problem with a single, official formula (not cask) please file this issue at Homebrew/homebrew-core: https://github.com/Homebrew/homebrew-core/issues/new/choose. If it's a `brew cask` problem please file this issue at https://github.com/Homebrew/homebrew-cask/issues/new/choose. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [ X ] ran `brew update` and can still reproduce the problem?
- [ X ] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [ X ] ran `brew config` and `brew doctor` and included their output with your issue?

Working brew config:

```
$ brew config
HOMEBREW_VERSION: 1.7.0
ORIGIN: https://github.com/Homebrew/brew
HEAD: 7e8fb9a0f8ab5e841763ec6c7fefa72a8462b594
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 1a0e4496243f6f8cd7c2ead65f7d9b0e640dedb9
Core tap last commit: 3 hours ago
HOMEBREW_PREFIX: /usr/local
CPU: quad-core 64-bit broadwell
Homebrew Ruby: 2.3.7 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/bin/ruby
Clang: 8.0 build 800
Git: 2.10.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.49.1 => /usr/bin/curl
Java: N/A
macOS: 10.12.1-x86_64
CLT: 8.2.0.0.1.1480973914
Xcode: N/A
XQuartz: 2.7.11 => /opt/X11
```

```
$ brew doctor
Your system is ready to brew.
```

Non-working brew config:

```
» brew config
HOMEBREW_VERSION: 1.7.0-42-g41680cd
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: 41680cd23915e8a1cca5f4945d3ec34d7dea3e82
Last commit: 2 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 1a0e4496243f6f8cd7c2ead65f7d9b0e640dedb9
Core tap last commit: 3 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_DEV_CMD_RUN: 1
CPU: octa-core 64-bit haswell
Homebrew Ruby: 2.3.7 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/bin/ruby
Clang: 9.0 build 900
Git: 2.14.2 => /usr/local/bin/git
Curl: 7.54.0 => /usr/bin/curl
Java: 1.8.0_45
macOS: 10.12.6-x86_64
CLT: 9.2.0.0.1.1510905681
Xcode: N/A
XQuartz: 2.7.11 => /opt/X11
```

```
» brew doctor
Your system is ready to brew.
```

To help us debug your issue please explain:

- What you were trying to do (and why)

Install a locally-produced package that contains a single Go binary at the top-level of the archive.

- What happened (include command output)

```
➜ ~❯ brew doctor
Your system is ready to brew.
➜ ~❯ brew upgrade gobinary-cli
==> Upgrading 1 outdated package, with result:
org/tap/gobinary-cli 0.32.1 -> 0.33.0
==> Upgrading org/tap/gobinary-cli
==> Downloading https://local.homebrew.tap/homebrew/gobinary-cli-darwin-v0.33.0.tar.gz
Already downloaded: /Users/morr/Library/Caches/Homebrew/gobinary-cli-0.33.0.tar.gz
Error: No such file or directory - gobinary
➜ ~❯
```

- What you expected to happen

The installation of the binary to complete.

- Step-by-step reproduction instructions (by running `brew` commands)

1. Produce a homebrew package that contains a file at the top level of the archive, e.g.:

```
~ $ tar -tvf gobinary-cli-darwin-v0.33.0.tar.gz
-rwxrwxr-x  0 root   root 30329808 Jul 17 15:12 gobinary
~ $
```

2. Generate and publish a homebrew formula for installing it, e.g. :

```
class GobinaryCli < Formula
  desc "gobinary-cli"
  homepage "https://local.git.ui/org/gobinary-cli"
  version "0.33.0"
  url "https://local.homebrew.tap/homebrew/gobinary-cli-darwin-v0.33.0.tar.gz"
  sha256 "459b3c52b96af66797907411f3ead0a4d43a9f74850f37a5ce601128485bd159"

  def install
    bin.install "gobinary"
  end
end
```

Using homebrew revision 41680cd23915e8a1cca5f4945d3ec34d7dea3e82, installation fails, as it appears the binary is never symlinked, then an exception is raised at:

https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/pathname.rb#L94

Using homebrew revision 7e8fb9a0f8ab5e841763ec6c7fefa72a8462b594, all is well.

Looking at the diff between these two revisions, it looks like some nested archive handling was added, perhaps it was due to this.

3. Produce a formula that contains the same binary in a folder, e.g:

```
~ $ tar -tvf gobinary-darwin-v0.33.0.tar.gz
drwxr-xr-x  0 root  root       0 Jul 17 15:12 gobinary-darwin/
-rwxr-xr-x  0 root  root 30329808 Jul 17 15:12 gobinary-darwin/gobinary
```

4. Attempt to install this package using the same formula, and it works in both revisions.

I'm not certain if installation of top-level binaries in the archive is supported or not, but given that it *used* to work, and is now seemingly broken, I thought it'd be worth reporting.

**Comment by reitermarkus** [comment]
This was probably also fixed by https://github.com/Homebrew/brew/pull/4511.

**Comment by adamwallred** [comment]
It does seem to be fixed after #4511, the unpack completes with:

```
» brew unpack gobinary-cli --verbose --debug
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/org/homebrew-tap/gobinary-cli.rb
==> Unpacking org/tap/gobinary-cli to: /Volumes/CaseSensitive/projects/org/repo/gobinary-cli-0.33.0
==> Downloading https://internal.brew.tap/homebrew/gobinary-cli-darwin-v0.33.0.tar.gz
Already downloaded: /Users/username/Library/Caches/Homebrew/gobinary-cli-0.33.0.tar.gz
==> Verifying gobinary-cli-0.33.0.tar.gz checksum
tar xf /Users/username/Library/Caches/Homebrew/gobinary-cli-0.33.0.tar.gz -C /var/folders/5g/r_p4k_hn74v_9f6lgl5k882r0000gp/T/d20180719-24264-1kd7q8l
```

The single go binary is present and linked. The install methods completes without error.

**Comment by reitermarkus** [comment]
Thank you for confirming, @adamwallred.
